File: Python/Evaluate_Reverse_Polish_Notation.py
#!/usr/bin/env python
# coding=utf-8

class Solution:
    # @param tokens, a list of a string
    # @return an integer
    def evalRPN(self, tokens):
        mstack = []
        for token in tokens:
            if token in ("+", "-", "*", "/"):
                b=mstack.pop()
                a=mstack.pop()
                if token == "+":
                    mstack.append(a+b)
                elif token == "-":
                    mstack.append(a-b)
                elif token == "*":
                    mstack.append(a*b)
                elif token == "/":
                    c=a/b
                    if c<0 and float(c) != float(a)/float(b):
                        c += 1
                    mstack.append(c)
            else:
                mstack.append(int(token))
        return mstack.pop()

File: java/Evaluate_Reverse_Polish_Notation.java
// Stack solution
// Time is O(n), space is O(n).

public class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<Integer>();
        int n = tokens.length,a;
        if (n == 0) return 0;
        for (int i = 0; i < n; ++i) {
            char op = tokens[i].charAt(0);
            if (tokens[i].length() == 1 && (op > '9' || op < '0')) {
                switch (op) {
                    case '+':
                        stack.push(stack.pop()+stack.pop());
                        break;
                    case '-':
                        a = stack.pop();
                        stack.push(stack.pop() - a);
                        break;
                    case '*':
                        stack.push(stack.pop()*stack.pop());
                        break;
                    case '/':
                        a = stack.pop();
                        stack.push(stack.pop()/a);
                        break;
                }
            }
            else {
                stack.push(Integer.parseInt(tokens[i]));
            }
        }
        return stack.pop();
    }
}
